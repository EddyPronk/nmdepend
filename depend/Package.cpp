// Copyright (c) 2004 E.J. Pronk and R.E.W. van Beusekom
//
// This file is part of Nmdepend.
// 
// Nmdepend is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// Nmdepend is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with Nmdepend; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#include "Package.h"

#include <boost/filesystem/convenience.hpp>
#include <boost/filesystem/operations.hpp>

#include <iostream>

namespace fs = boost::filesystem;

Package::PackageRegistry Package::s_PackageRegistry;

Package::Package(const fs::path& p)
{
  m_Package = s_PackageRegistry.find(p.string()/*.leaf()*/);

  if(m_Package == s_PackageRegistry.end())
  {
    std::pair<PackageRegistry::iterator, bool> ret =
      s_PackageRegistry.insert(
	std::make_pair(p.string()/*.leaf()*/, PackageContent())
      );
    if(ret.second)
    {
      m_Package = ret.first;
    }
  }

  ++m_Package->second.refcount;

  fs::directory_iterator end;
  for(fs::directory_iterator it(p); it != end; ++it)
  {
    if(fs::is_directory(*it))
    {
      Package pack(*it);
      if(pack)
      {
	m_Package->second.package.insert(pack);
      }
    }
    else
    {
      if(fs::extension(*it) == ".o")
      {
	Object o(*it);
        m_Package->second.object.insert(o);
      }
    }
  }
  if(m_Package->second.object.empty() && m_Package->second.package.empty())
  {
    s_PackageRegistry.erase(m_Package);
    m_Package = s_PackageRegistry.end();
  }
}
Package::Package(const Package& p):
  m_Package(p.m_Package)
{
  ++m_Package->second.refcount;
}
Package::~Package()
{
  if(--m_Package->second.refcount == 0)
  {
    s_PackageRegistry.erase(m_Package);
  }
}
bool Package::operator < (const Package& p) const
{
  return m_Package->first < p.m_Package->first;
}
bool Package::operator == (const Package& p) const
{
  return m_Package->first == p.m_Package->first;
}
Package::operator bool() const
{
  return m_Package != s_PackageRegistry.end();
}
const std::string& Package::Name() const
{
  return m_Package->first;
}
const std::set<Object>& Package::Objects() const
{
  return m_Package->second.object;
}
const std::set<Package>& Package::Packages() const
{
  return m_Package->second.package;
}
bool Package::Depend(const Object& o) const
{
  bool depends(false);

  for(std::set<Object>::const_iterator it = m_Package->second.object.begin();
      it != m_Package->second.object.end(); ++it)
  {
    if(it->Depend(o))
    {
      //direct dependency
      depends = true;
    }
  }
  
  for(std::set<Package>::const_iterator it = m_Package->second.package.begin();
      it != m_Package->second.package.end(); ++it)
  {
     it->Depend(o);
  }

  return depends;
}
bool Package::Depend(const Package& p) const
{
  bool depends(false);

  std::string cluster(fs::basename(fs::path(m_Package->first).leaf()));
  if(cluster.empty())
  {
    cluster = fs::basename(fs::absolute(m_Package->first).filename());
  }

  std::cout << "subgraph cluster_"
            << cluster
            << "{" << std::endl
            << "label=" << cluster
            << ';' << std::endl;

  for(std::set<Object>::const_iterator it = m_Package->second.object.begin();
      it != m_Package->second.object.end(); ++it)
  {
    std::cout << '"' << fs::basename(fs::path(it->Name()).leaf()) << "\";" << std::endl;
  }

  for(std::set<Package>::const_iterator it = m_Package->second.package.begin();
      it != m_Package->second.package.end(); ++it)
  {
    it->Depend(p);
  }

  std::cout << "}" << std::endl;

  return depends;
}

bool Package::PackDepend(const Package& p) const
{
  bool depends(false);

  Depend(p);

  for(std::set<Package>::const_iterator it = m_Package->second.package.begin();
      it != m_Package->second.package.end(); ++it)
  {
    p.PackDepend(*it);
  }

  for(std::set<Object>::const_iterator it = m_Package->second.object.begin();
      it != m_Package->second.object.end(); ++it)
  {
    if(it->Depend(p))
    {
      //direct dependency
      depends = true;
    }
  }

  if(depends && Name() != p.Name())
  {
    std::clog << '"' << fs::basename(fs::path(Name()).leaf()) << "\" -> \""
              << fs::basename(fs::path(p.Name()).leaf()) << "\";" << std::endl;
  }
  return depends;
}
