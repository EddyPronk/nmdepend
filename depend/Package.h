// Copyright (c) 2004 E.J. Pronk and R.E.W. van Beusekom
//
// This file is part of Nmdepend.
// 
// Nmdepend is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// Nmdepend is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with Nmdepend; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#ifndef PACKAGE_H
#define PACKAGE_H

#include <boost/filesystem/path.hpp>

#include "Object.h"

class Package
{
  struct PackageContent
  {
    PackageContent(): refcount(0){}
    unsigned int refcount;
    std::set<Object> object;
    std::set<Package> package;
  };
public:
  typedef std::string Name_t;
  typedef std::map<Package::Name_t, Package::PackageContent > PackageRegistry;
private:
  static PackageRegistry s_PackageRegistry;
  PackageRegistry::iterator m_Package;
public:
  Package(const boost::filesystem::path& p);
  Package(const Package& p);
  ~Package();
  bool operator < (const Package& p) const;
  bool operator == (const Package& p) const;
  operator bool() const;
  const std::string& Name() const;
  const std::set<Object>& Objects() const;
  const std::set<Package>& Packages() const;
  bool Depend(const Object& o) const;
  bool Depend(const Package& p) const;
  bool PackDepend(const Package& p) const;
};


#endif
