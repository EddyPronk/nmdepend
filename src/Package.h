#ifndef PACKAGE_H
#define PACKAGE_H

#include <string>
#include <vector>
#include <set>
#include <map>

// Package does not implement Provides/Requires containers.

class Package
{
public:
   typedef std::string Name_t;
   typedef std::set<Package*> Index_t;
   typedef std::vector<Package*> SubPackageList_t;
   typedef std::map<Name_t, Package*> PackageRegistry_t;


   Package(const Name_t& name);

   void SetParent(Package& parent);
   //Package& Parent();
   void Imports();
   void Imports(SubPackageList_t& list);
   Name_t& Name();

   // todo Rename/remove this
   Name_t Name2();

   // Two types of relationships are being stored.
   // Import/Export represent dependencies between packages on same level.
   // Requires/Provides is the same as undefined/defined for objectfiles.
   // These names are choosen to abstract from objectfiles.
   //
   // Import/Export has different semantics as in UML.
   //
   //   Provides is a 'has' relationship
   //   Requires is a list o

   virtual void AddRequires(Package* p);
   virtual void AddProvides(Package* p);
   virtual void AddImport(Package* p);
   virtual void AddExport(Package* p);

private:

   // For storing inter package dependencies
   std::set<Package*> m_Imports;
   std::set<Package*> m_Exports;

   Name_t m_Name;

protected:
   Package* m_Parent;
};

#endif
