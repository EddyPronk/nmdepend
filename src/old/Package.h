#ifndef PACKAGE_H
#define PACKAGE_H

class Package : public IPackage
{
public:
   //typedef std::vector<IPa
   Package(const IPackage::Name_t& name)
      : IPackage(name)
   {
   }

private:
   Index_t m_Imports;
   Index_t m_Exports;
};

#endif
