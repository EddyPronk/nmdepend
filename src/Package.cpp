// Copyright (c) 2004 E.J. Pronk and R.E.W. van Beusekom
//
// This file is part of Nmdepend.
// 
// Nmdepend is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// Nmdepend is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with Nmdepend; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#ifdef _MSC_VER
#pragma warning( disable : 4786)
#endif

#include "Package.h"
#include <iostream>

Package::Package(Callback<Package>& callback, const std::string& name)
 : Entity(name)
 , m_Parent(0)
 , m_Callback(callback)
{
}

std::string Package::Name2()
{
   std::string name = Name();

  //todo configure extension
  //in the meantime using rfind(".o") should work for *.obj too
  return name.substr(0, name.rfind(".o"));
}

/*void Package::SetParent(Package& parent)
{
   m_Parent = &parent;
}*/

void Package::AddImport(Package* p)
{
   if (p != this)
   {
      std::cout << "Package::AddImport " << p->Name() << " to " << Name() << std::endl;
      m_Imports.insert(p);
   }

   if (m_Parent && p->m_Parent)
   {
      m_Parent->AddImport(p->m_Parent);
   }
}

void Package::AddExport(Package* p)
{
   if (p != this)
   {
      std::cout << "Package::AddExport " << p->Name() << " to " << Name() << std::endl;
      m_Exports.insert(p);
   }

   if (m_Parent && p->m_Parent)
   {
      m_Parent->AddExport(p->m_Parent);
   }
}

void Package::Imports(SubPackageList_t& list)
{
   list.clear();
   for(Index_t::iterator pos = m_Imports.begin()
      ; pos != m_Imports.end(); ++pos)
   {
      list.push_back(*pos);
      std::cout << Name() << " -> " << (*pos)->Name() << std::endl;
   }
}

/*void Package::Provides(SubPackageList_t& list)
{
   std::cout << "Provides" << std::endl;
}*/

void Package::Link()
{
}

bool Package::Depend(const Package& package) const
{
  return true; 
}

void Package::Link(Package* rsh)
{
  m_Callback(*this, *rsh);
}

