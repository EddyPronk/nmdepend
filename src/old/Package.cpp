#include "Package.cpp"

Package::AddImport(IPackage& p)
{
   m_Imports.push_back(p);
}