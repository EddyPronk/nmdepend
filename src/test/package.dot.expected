digraph G {
0[label="app"];
1[label="sub1"];
2[label="sub2"];
0->1 ;
1->2 ;
}
